In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [22]:
sales = pd.read_excel("sales.xlsx")

sales.head()

,year,release_date,title,genre,international_box_office,domestic_box_office,worldwide_box_office,production_budget,Unnamed: 8,opening_weekend,theatre_count,avg run per theatre,runtime,keywords,creative_type,url
0,2000,January 1st,Bakha Satang,Drama,76576.0,NaN,76576.0,NaN,NaN,NaN,NaN,NaN,129.0,NaN,Contemporary Fiction,https://www.the-numbers.com/movie/Bakha-Satang...
1,2001,January 12th,Antitrust,Thriller/Suspense,6900000.0,10965209.0,17865209.0,30000000.0,NaN,5486209.0,2433.0,3.1,NaN,NaN,Contemporary Fiction,https://www.the-numbers.com/movie/Antitrust
2,2000,January 28th,Santitos,NaN,NaN,378562.0,NaN,NaN,NaN,NaN,NaN,NaN,105.0,NaN,NaN,https://www.the-numbers.com/movie/Santitos
3,2002,2002 (Wide) by,Frank McKlusky C.I.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,https://www.the-numbers.com/movie/Frank-McKlus...
4,2002,January 25th,A Walk to Remember,Drama,4833792.0,41227069.0,46060861.0,11000000.0,NaN,12177488.0,2411.0,5.3,NaN,Coming of Age,Contemporary Fiction,https://www.the-numbers.com/movie/Walk-to-Reme...


In [23]:
sales.info()
sales.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 30612 entries, 0 to 30611
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   year                      30612 non-null  int64  
 1   release_date              30612 non-null  str    
 2   title                     30604 non-null  object 
 3   genre                     28908 non-null  str    
 4   international_box_office  21575 non-null  float64
 5   domestic_box_office       11884 non-null  float64
 6   worldwide_box_office      21575 non-null  float64
 7   production_budget         4480 non-null   float64
 8   Unnamed: 8                0 non-null      float64
 9   opening_weekend           10929 non-null  float64
 10  theatre_count             10963 non-null  float64
 11  avg run per theatre       10952 non-null  float64
 12  runtime                   24559 non-null  float64
 13  keywords                  12517 non-null  object 
 14  creative_type    

year                            0
release_date                    0
title                           8
genre                        1704
international_box_office     9037
domestic_box_office         18728
worldwide_box_office         9037
production_budget           26132
Unnamed: 8                  30612
opening_weekend             19683
theatre_count               19649
avg run per theatre         19660
runtime                      6053
keywords                    18095
creative_type                3945
url                             0
dtype: int64

In [24]:
sales = sales.drop(columns=['Unnamed: 8'])

In the overview I saw that the columm Unnamed: 8 had 0 non-null values, this is clearly an empty columm and therefore it can ben removed. 

In [25]:
sales['release_date'].unique()

<StringArray>
[             'January 1st',             'January 12th',
             'January 28th',          '2002 (Wide) by ',
             'January 25th',                'June 21st',
                 'May 10th',              'January 2nd',
                 'June 7th',             'January 18th',
 ...
           '2003 (Limited)',              '2002 (Wide)',
            '2020 (Wide) (',      '2020 (Canceled) by ',
 'Fall 2020 (Canceled) by ',          '2022 (Wide) by ',
            '2021 (Wide) (',                   'Summer',
      '2022 (Canceled) by ',      '2021 (Canceled) by ']
Length: 392, dtype: str

Here you can see that there is a widespread of versions that the release dat can have. 

In [26]:
sales['production_budget'].unique()

array([          nan, 3.0000000e+07, 1.1000000e+07, 2.2000000e+07,
       1.0000000e+07, 4.0000000e+07, 8.0000000e+05, 7.0000000e+07,
       3.4800000e+07, 5.0000000e+07, 1.3000000e+07, 1.0000000e+06,
       9.6000000e+06, 1.0000000e+08, 3.0600000e+08, 1.3500000e+08,
       6.0000000e+06, 1.0300000e+07, 9.0000000e+06, 4.5000000e+07,
       6.0000000e+07, 4.2000000e+07, 1.3000000e+08, 6.5000000e+05,
       7.8000000e+06, 3.5000000e+07, 1.2000000e+07, 8.5000000e+06,
       2.5000000e+07, 1.0200000e+08, 1.9800000e+07, 9.0000000e+07,
       1.5000000e+07, 6.8000000e+07, 1.5700000e+07, 2.6200000e+08,
       8.4000000e+07, 3.0000000e+05, 1.6000000e+08, 1.1100000e+08,
       4.0000000e+06, 1.9500000e+07, 4.5000000e+06, 2.0000000e+07,
       6.5000000e+07, 1.2500000e+08, 1.7500000e+08, 9.9000000e+07,
       7.0000000e+06, 4.6500000e+07, 4.8000000e+07, 2.3000000e+07,
       1.1000000e+08, 9.5000000e+06, 3.8000000e+07, 2.4000000e+07,
       2.0000000e+08, 3.6000000e+07, 1.5000000e+08, 1.2000000e

In the overview I saw that production_budget had quite some null-values 26132 of the 30612, therefore i wanted to double check this if I could get a clearer view on this but I could not. For the null-values I can't take a median or mean, just because there are to many values missing, I will stike with the NaN values. 

In [27]:
sales['title_clean'] = sales['title'].str.strip().str.lower()
sales.head()

,year,release_date,title,genre,international_box_office,domestic_box_office,worldwide_box_office,production_budget,opening_weekend,theatre_count,avg run per theatre,runtime,keywords,creative_type,url,title_clean
0,2000,January 1st,Bakha Satang,Drama,76576.0,NaN,76576.0,NaN,NaN,NaN,NaN,129.0,NaN,Contemporary Fiction,https://www.the-numbers.com/movie/Bakha-Satang...,bakha satang
1,2001,January 12th,Antitrust,Thriller/Suspense,6900000.0,10965209.0,17865209.0,30000000.0,5486209.0,2433.0,3.1,NaN,NaN,Contemporary Fiction,https://www.the-numbers.com/movie/Antitrust,antitrust
2,2000,January 28th,Santitos,NaN,NaN,378562.0,NaN,NaN,NaN,NaN,NaN,105.0,NaN,NaN,https://www.the-numbers.com/movie/Santitos,santitos
3,2002,2002 (Wide) by,Frank McKlusky C.I.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,https://www.the-numbers.com/movie/Frank-McKlus...,frank mcklusky c.i.
4,2002,January 25th,A Walk to Remember,Drama,4833792.0,41227069.0,46060861.0,11000000.0,12177488.0,2411.0,5.3,NaN,Coming of Age,Contemporary Fiction,https://www.the-numbers.com/movie/Walk-to-Reme...,a walk to remember


Here I normalised the title columm, i made sure that there were only lower cases in the titles and added that to the new colomm title_clean

In [28]:
sales['match_key'] = sales['title_clean'] + '_' + sales['year'].astype(str)
sales['match_key'].head()

0           bakha satang_2000
1              antitrust_2001
2               santitos_2000
3    frank mcklusky c.i._2002
4     a walk to remember_2002
Name: match_key, dtype: str

Here I made a match_key so that I eventually can make the new id where I can math on. This because I don't want any letters in my id (primary keys)

In [29]:
sales['match_key'].duplicated().sum()

np.int64(107)

I was checking for duplicates in the new colomm and it gave me three, so now im gonna check which ones they are.

In [30]:
print(sales['match_key'].isnull().sum())

78


Here I checked all the nullvalues in Match_key to see where the fault is made.

In [31]:
duplicates = sales[sales['match_key'].notnull() & sales['match_key'].duplicated(keep=False)].sort_values('match_key')
duplicates[['title', 'match_key']]

,title,match_key
191,A Christmas Carol,a christmas carol_2018
338,A Christmas Carol,a christmas carol_2018
27120,Adam,adam_2020
25968,Adam,adam_2020
11826,Aftermath,aftermath_2014
8567,Aftermath,aftermath_2014
9818,Best F,best f_2018
6773,Best F,best f_2018
2330,Blue,blue_2017
2331,Blue,blue_2017


In [32]:
colliding_keys = sales['match_key'][sales['match_key'].duplicated(keep=False)]

sales.loc[sales['match_key'].isin(colliding_keys), 'match_key'] = (
    sales.loc[sales['match_key'].isin(colliding_keys), 'match_key'] 
    + '_' + sales.loc[sales['match_key'].isin(colliding_keys), 'runtime'].astype(str)
)

So I'm gonna leave this duplicates, but I need to add something to the name otherwise they will all get matched to the other id's. Therefore I added the runtime, this because in the metaclean data runtime comes up to, this colomm is the most unique and all the runtimes differ from eachother so that is the best one to match on. 

In [33]:
sales['match_key'].duplicated().sum()

np.int64(82)

So now i added runtime to the name and there are still 82 duplicate values. 

In [34]:
print(sales['match_key'].duplicated(keep=False).sum())
print(sales['match_key'][sales['match_key'].duplicated(keep=False)].nunique())

83
0


I did this check to see how many rows there are who are duplicate, and the second one is to see if there are any unique keys who are duplicate. 

In [35]:
print(sales['match_key'].isnull().sum())

83


In [36]:
print(sales[sales['match_key'].isnull()][['title', 'title_clean', 'year']])

                     title        title_clean  year
14                     120                NaN  2008
191      A Christmas Carol  a christmas carol  2018
566                      2                NaN  2018
1248                   808                NaN  2016
1571   2016-11-08 00:00:00                NaN  2017
...                    ...                ...   ...
29083                 1921                NaN  2021
29389                  NaN                NaN  2020
29395                 2050                NaN  2020
29863                 1986                NaN  2021
30552                 1982                NaN  2021

[83 rows x 3 columns]


In [37]:
sales['title'] = sales['title'].astype(str)
sales['title'] = sales['title'].str.replace(r'\.0$', '', regex=True)

sales['title_clean'] = sales['title'].str.strip().str.lower()
sales['match_key'] = sales['title_clean'] + '_' + sales['year'].astype(str)

colliding_keys = sales['match_key'][sales['match_key'].duplicated(keep=False)]
runtime_str = sales.loc[sales['match_key'].isin(colliding_keys), 'runtime'].fillna('onbekend').astype(str)

sales.loc[sales['match_key'].isin(colliding_keys), 'match_key'] = (
    sales.loc[sales['match_key'].isin(colliding_keys), 'match_key'] + '_' + runtime_str
)

print('match_key null:', sales['match_key'].isnull().sum())
print('duplicated():', sales['match_key'].duplicated().sum())

match_key null: 8
duplicated(): 7


So before I could even build a proper match_key, I noticed that some titles were not being read as text at all. Excel had turned some movie titles into actual numbers or even dates (like "1985" or "2016-11-08"), so when I tried to clean the title column with .str functions, those rows just turned into empty values instead of a proper title. I fixed this by forcing the whole title column to text first with astype(str), and then removed the extra ".0" that showed up on the ones that were read as numbers.

After that I rebuilt title_clean and match_key like before, using title and year. Some of these match_keys were still duplicate, so just like before I need to add something extra to make them unique, otherwise they'll all get matched to the wrong movie later. I used runtime again since that's the most unique column both datasets have and it almost always differs between two movies that share the same title and year.

One thing I ran into is that when runtime was missing for one of the duplicate movies, just adding it straight away actually turned the whole match_key into an empty value instead of keeping it unique, because pandas doesn't turn a missing number into the text "unknown" by itself, it stays empty even after converting to text. So I first filled the missing runtimes with the word "onbekend" before adding it to the match_key, that way even the movies with no runtime info keep a working, unique match_key instead of disappearing into empty values.

In [38]:
sales.head()

,year,release_date,title,genre,international_box_office,domestic_box_office,worldwide_box_office,production_budget,opening_weekend,theatre_count,avg run per theatre,runtime,keywords,creative_type,url,title_clean,match_key
0,2000,January 1st,Bakha Satang,Drama,76576.0,NaN,76576.0,NaN,NaN,NaN,NaN,129.0,NaN,Contemporary Fiction,https://www.the-numbers.com/movie/Bakha-Satang...,bakha satang,bakha satang_2000
1,2001,January 12th,Antitrust,Thriller/Suspense,6900000.0,10965209.0,17865209.0,30000000.0,5486209.0,2433.0,3.1,NaN,NaN,Contemporary Fiction,https://www.the-numbers.com/movie/Antitrust,antitrust,antitrust_2001
2,2000,January 28th,Santitos,NaN,NaN,378562.0,NaN,NaN,NaN,NaN,NaN,105.0,NaN,NaN,https://www.the-numbers.com/movie/Santitos,santitos,santitos_2000
3,2002,2002 (Wide) by,Frank McKlusky C.I.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,https://www.the-numbers.com/movie/Frank-McKlus...,frank mcklusky c.i.,frank mcklusky c.i._2002
4,2002,January 25th,A Walk to Remember,Drama,4833792.0,41227069.0,46060861.0,11000000.0,12177488.0,2411.0,5.3,NaN,Coming of Age,Contemporary Fiction,https://www.the-numbers.com/movie/Walk-to-Reme...,a walk to remember,a walk to remember_2002


In [39]:
sales.insert(0, "sales_id", range(1, len(sales) + 1))

print(sales[["sales_id", "title", "match_key"]].head(5))

   sales_id                title                 match_key
0         1         Bakha Satang         bakha satang_2000
1         2            Antitrust            antitrust_2001
2         3             Santitos             santitos_2000
3         4  Frank McKlusky C.I.  frank mcklusky c.i._2002
4         5   A Walk to Remember   a walk to remember_2002


Here I made a sales_id for the primary key to connect all the tables with eachother 

In [40]:
sales.to_csv('sales_cleaned.csv', index=False)

In [41]:
sales =  pd.read_csv('sales_cleaned.csv')

sales_table = sales.rename(columns={"avg run per theatre": "avg_run_per_theatre"})

sales_table = sales_table[["sales_id", "production_budget", "international_box_office", "domestic_box_office", "opening_weekend", "theatre_count", "avg_run_per_theatre", "creative_type"]]

sales_table["theatre_count"] = sales_table["theatre_count"].astype("Int64")

sales_table.to_csv('sales_table.csv', index=False)

print(sales_table.shape)
print(sales_table["sales_id"].is_unique)
print(sales_table.columns.tolist())
print(sales["title"].isna().sum(), "sales rows without a title")
print(sales_table["production_budget"].notna().sum(), "rows with a production_budget")

(30612, 8)
True
['sales_id', 'production_budget', 'international_box_office', 'domestic_box_office', 'opening_weekend', 'theatre_count', 'avg_run_per_theatre', 'creative_type']
8 sales rows without a title
4480 rows with a production_budget


Here I created the SALES table. This is the cleaned sales data, but then reduced to only the columns that are in the ERD, so sales_id, production_budget, international_box_office, domestic_box_office, opening_weekend, theatre_count, avg_run_per_theatre and creative_type. All the other columns (like year, release_date, title, genre, worldwide_box_office, runtime, keywords, url, title_clean and match_key) are dropped. The title, year and match_key were only needed to link the sales data to the movies, and that link now lives in the MOVIE_SALES table. Because I needed the match_key for that, I kept the full sales dataframe untouched and made the SALES table as a separate dataframe.

I also renamed avg run per theatre to avg_run_per_theatre, so it has the same name as in the ERD and I don't need quotes around the column name in PostgreSQL later. The theatre_count was a decimal number, so I converted it to a whole number, because the ERD defines it as an integer. I did not remove any rows. All 30,612 sales rows stay in the table, also the ones that are not linked to a movie in MOVIE_SALES, and I checked that every sales_id is unique. A lot of values in this table are empty, for example the production_budget is only filled for 4,480 rows. That is normal for this data, but it is good to keep in mind for the analysis.

One thing I noticed is that 8 sales rows have no title, and therefore no match_key. These are not empty rows: for example 500 Days of Summer is one of them. The title probably got lost in the cleaning because it starts with a bracket. This means these movies can't be linked to the MOVIE table. With only 8 out of 30,612 rows I decided to leave it like this. In the end I saved the result as sales_table.csv.